[DCP][state_dict] DCP state_dict cannot correctly find FQN when the leaf module is wrapped by FSDP

fegin opened the pull request:
Summary: The original logic has an incorrect assumption that there is at one object name left when traversing the module tree. This is not correct when the leaf module is wrapped by FSDP.

Test Plan: CI

Differential Revision: D52049293




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

wz337 commented on test/distributed/checkpoint/test_state_dict.py:
```diff
@@ -181,7 +191,7 @@ def test_fsdp(self) -> None:
                 "use_orig_params": [True, False],
                 "use_composable": [True, False],
-                "use_dtensor": [True, False],
+                "use_dtensor": [True,False],
```
This seems unintentional?

wz337 reviewed (approved):
LGTM!

fegin committed:
[DCP][state_dict] DCP state_dict cannot correctly find FQN when the leaf module is wrapped by FSDP (#115592)

Summary:

The original logic has an incorrect assumption that there is at one object name left when traversing the module tree. This is not correct when the leaf module is wrapped by FSDP.

Test Plan: CI

Reviewed By: wz337

Differential Revision: D52049293
